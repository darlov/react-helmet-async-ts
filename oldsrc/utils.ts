import {TAG_NAMES, TAG_PROPERTIES, ATTRIBUTE_NAMES, SEO_PRIORITY_TAGS} from './constants';
import {
  HelmetProps,
  BaseProps,
  AttributeNamesMap,
  TagPropsType,
  IStateType,
  MetaProps,
  LinkProps,
  ScriptProps
} from "./types";

enum HELMET_PROPS {
  DEFAULT_TITLE = 'defaultTitle',
  DEFER = 'defer',
  ENCODE_SPECIAL_CHARACTERS = 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE = 'onChangeClientState',
  TITLE_TEMPLATE = 'titleTemplate',
  PRIORITIZE_SEO_TAGS = 'prioritizeSeoTags'
}

const getInnermostProperty = <T extends keyof HelmetProps>(propsList: HelmetProps[], property: T) => {
  for (let i = propsList.length - 1; i >= 0; i -= 1) {
    const props = propsList[i];

    if (Object.hasOwn(props, property)) {
      return props[property];
    }
  }

  return undefined;
};

const getTitleFromPropsList = (propsList: HelmetProps[]) => {
  let innermostTitle = getInnermostProperty(propsList, TAG_NAMES.TITLE);
  const innermostTemplate = getInnermostProperty(propsList, HELMET_PROPS.TITLE_TEMPLATE);
  if (Array.isArray(innermostTitle)) {
    innermostTitle = innermostTitle.join('');
  }
  if (innermostTemplate && innermostTitle) {
    // use function arg to avoid need to escape $ characters
    return innermostTemplate.replace(/%s/g, innermostTitle);
  }

  const innermostDefaultTitle = getInnermostProperty(propsList, HELMET_PROPS.DEFAULT_TITLE);

  return innermostTitle || innermostDefaultTitle || undefined;
};

const getOnChangeClientState = (propsList: HelmetProps[]) =>
  getInnermostProperty(propsList, HELMET_PROPS.ON_CHANGE_CLIENT_STATE) || (() => {
  });

const getAttributesFromPropsList = <T extends ATTRIBUTE_NAMES>(tagType: T, propsList: HelmetProps[]) =>
  propsList
    .map(props => props[tagType])
    .filter((tag): tag is AttributeNamesMap[T] => !!tag)
    .reduce<AttributeNamesMap[T]>((tagAttrs, current) => ({...tagAttrs, ...current}), {});

const getBaseTagFromPropsList = (primaryAttributes: TAG_PROPERTIES[], propsList: HelmetProps[]) =>
  propsList
    .map(props => props.base)
    .filter((tag): tag is BaseProps => !!tag)
    .reverse()
    .reduce<BaseProps[]>((innermostBaseTag, tag) => {
      if (!innermostBaseTag.length) {
        const keys = Object.keys(tag);

        for (let i = 0; i < keys.length; i += 1) {
          const attributeKey = keys[i];
          const lowerCaseAttributeKey = attributeKey.toLowerCase();

          if (primaryAttributes.some(m => m === lowerCaseAttributeKey) && tag[lowerCaseAttributeKey as keyof BaseProps]) {
            return innermostBaseTag.concat(tag);
          }
        }
      }

      return innermostBaseTag;
    }, []);

// eslint-disable-next-line no-console
const warn = (msg: string) => console && typeof console.warn === 'function' && console.warn(msg);

type Flatten<T> = T extends any[] ? T[number] : T;

const getTagsFromPropsList = <T extends keyof TagPropsType,
  TValidTags extends Required<TagPropsType>,
  TTags = TagPropsType[T],
  TElement = Flatten<TTags>>(tagName: T, primaryAttributes: TAG_PROPERTIES[], propsList: TagPropsType[]) => {
  // Calculate list of tags, giving priority innermost component (end of the propslist)
  const approvedSeenTags: { [key in keyof TElement]: any; } = {} as TElement;

  return propsList
    .filter((props): props is TValidTags => {
      if (Array.isArray(props[tagName])) {
        return true;
      }
      if (typeof props[tagName] !== 'undefined') {
        warn(
          `Helmet: ${ tagName } should be of type "Array". Instead found type "${ typeof props[
            tagName
            ] }"`
        );
      }
      return false;
    })
    .map((props) => props[tagName])
    .reverse()
    .reduce((approvedTags, instanceTags) => {
      const instanceSeenTags: { [key in keyof TElement]: any; } = {} as TElement;
      
      instanceTags
        .filter(tag => {
          let primaryAttributeKey: string | undefined = undefined;
          const keys = Object.keys(tag);
          for (let i = 0; i < keys.length; i += 1) {
            const attributeKey = keys[i];
            const lowerCaseAttributeKey = attributeKey.toLowerCase();

            // Special rule with link tags, since rel and href are both primary tags, rel takes priority
            if (
              primaryAttributes.some(m => m === lowerCaseAttributeKey) &&
              !(
                primaryAttributeKey === TAG_PROPERTIES.REL &&
                tag[primaryAttributeKey as keyof typeof tag].toLowerCase() === 'canonical'
              ) &&
              !(
                lowerCaseAttributeKey === TAG_PROPERTIES.REL &&
                tag[lowerCaseAttributeKey as keyof typeof tag].toLowerCase() === 'stylesheet'
              )
            ) {
              primaryAttributeKey = lowerCaseAttributeKey;
            }
            // Special case for innerHTML which doesn't work lowercased
            if (
              primaryAttributes.some(m => m === attributeKey) &&
              (attributeKey === TAG_PROPERTIES.INNER_HTML ||
                attributeKey === TAG_PROPERTIES.CSS_TEXT ||
                attributeKey === TAG_PROPERTIES.ITEM_PROP)
            ) {
              primaryAttributeKey = attributeKey;
            }
          }

          if (!primaryAttributeKey || !tag[primaryAttributeKey as keyof typeof tag]) {
            return false;
          }

          const value: string = tag[primaryAttributeKey as keyof typeof tag].toLowerCase();

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement]) {
            approvedSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!instanceSeenTags[primaryAttributeKey as keyof TElement]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement][value]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement][value] = true;
            return true;
          }

          return false;
        }, [])
        .reverse()
        .forEach(tag => approvedTags.push(tag as TElement));

      // Update seen tags with tags from this instance
      const keys = Object.keys(instanceSeenTags);
      for (let i = 0; i < keys.length; i += 1) {
        const attributeKey = keys[i];
        const tagUnion = {
          ...approvedSeenTags[attributeKey as keyof TElement],
          ...instanceSeenTags[attributeKey as keyof TElement],
        };

        approvedSeenTags[attributeKey as keyof TElement] = tagUnion;
      }

      return approvedTags;
    }, [] as TElement[])
    .reverse();
};

const getAnyTrueFromPropsList = <T extends keyof HelmetProps>(propsList: HelmetProps[], checkedTag: T) => {
  if (Array.isArray(propsList) && propsList.length) {
    for (let index = 0; index < propsList.length; index += 1) {
      const prop = propsList[index];
      if (prop[checkedTag]) {
        return true;
      }
    }
  }
  return false;
};
const reducePropsToState = (propsList: HelmetProps[]): IStateType => ({
  baseTag: getBaseTagFromPropsList([TAG_PROPERTIES.HREF], propsList),
  bodyAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.BODY, propsList),
  defer: getInnermostProperty(propsList, HELMET_PROPS.DEFER),
  encode: getInnermostProperty(propsList, HELMET_PROPS.ENCODE_SPECIAL_CHARACTERS),
  htmlAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.HTML, propsList),
  linkTags: getTagsFromPropsList(
    TAG_NAMES.LINK,
    [TAG_PROPERTIES.REL, TAG_PROPERTIES.HREF],
    propsList
  ),
  metaTags: getTagsFromPropsList(
    TAG_NAMES.META,
    [
      TAG_PROPERTIES.NAME,
      TAG_PROPERTIES.CHARSET,
      TAG_PROPERTIES.HTTPEQUIV,
      TAG_PROPERTIES.PROPERTY,
      TAG_PROPERTIES.ITEM_PROP,
    ],
    propsList
  ),
  noscriptTags: getTagsFromPropsList(TAG_NAMES.NOSCRIPT, [TAG_PROPERTIES.INNER_HTML], propsList),
  onChangeClientState: getOnChangeClientState(propsList),
  scriptTags: getTagsFromPropsList(
    TAG_NAMES.SCRIPT,
    [TAG_PROPERTIES.SRC, TAG_PROPERTIES.INNER_HTML],
    propsList
  ),
  styleTags: getTagsFromPropsList(TAG_NAMES.STYLE, [TAG_PROPERTIES.CSS_TEXT], propsList),
  title: getTitleFromPropsList(propsList),
  titleAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.TITLE, propsList),
  prioritizeSeoTags: getAnyTrueFromPropsList(propsList, HELMET_PROPS.PRIORITIZE_SEO_TAGS),
});

export const flattenArray = (possibleArray: any) =>
  Array.isArray(possibleArray) ? possibleArray.join('') : possibleArray;

export { reducePropsToState };

const checkIfPropsMatch = (props: any, toMatch: any) => {
  const keys = Object.keys(props);
  for (let i = 0; i < keys.length; i += 1) {
    // e.g. if rel exists in the list of allowed props [amphtml, alternate, etc]
    if (toMatch[keys[i]] && toMatch[keys[i]].includes(props[keys[i]])) {
      return true;
    }
  }
  return false;
};

export type Prioritizer =
  { element: MetaProps, propsToMatch: typeof SEO_PRIORITY_TAGS["meta"] }
  | { element: LinkProps, propsToMatch: typeof SEO_PRIORITY_TAGS["link"] }
  | { element: ScriptProps, propsToMatch: typeof SEO_PRIORITY_TAGS["script"] };

export const prioritizer = <TElement extends Prioritizer["element"], TMatch extends Prioritizer["propsToMatch"]>(
  elements: TElement[],
  propsToMatch: TMatch) => {

  const result: { priority: TElement[], default: TElement[] } = {priority: [], default: []};

  if (Array.isArray(elements)) {

    for (const element of elements) {
      if (checkIfPropsMatch(element, propsToMatch)) {
        result.priority.push(element);
      } else {
        result.default.push(element);
      }
    }
  }
  return  result;
};

export const without = (obj: any, key: any) => {
  return {
    ...obj,
    [key]: undefined,
  };
}
