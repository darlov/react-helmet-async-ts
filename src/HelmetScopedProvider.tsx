import { createContext, useContext } from "react";
import {
  BaseProps,
  BodyProps, HtmlProps, IHelmetInstanceState,
  LinkProps,
  MetaProps,
  NoscriptProps,
  ScriptProps,
  StyleProps,
  TagProps,
  TitleProps, UpdateInstanceCallback
} from "./types";

export interface ITagsActions<T extends TagProps> {
  add: (tag: T) => void;
  remove: (tag: T) => void;
}

export const createActionsData = <
  T extends TagProps,
>(
  instance: IHelmetInstanceState,
  propName: keyof Omit<IHelmetInstanceState, "id" >,
  context:{addItem: UpdateInstanceCallback, removeItem: UpdateInstanceCallback }): ITagsActions<T> => {
  return {
    add: (value) => context.addItem(instance, propName, value),
    remove: (value) => context.removeItem(instance, propName, value)
  }
}

export interface IHelmetScopedContextData {
  titleActions: ITagsActions<TitleProps>;
  metaActions: ITagsActions<MetaProps>;
  styleActions: ITagsActions<StyleProps>;
  scriptActions: ITagsActions<ScriptProps>;
  linkActions: ITagsActions<LinkProps>;
  noscriptActions: ITagsActions<NoscriptProps>;
  baseActions: ITagsActions<BaseProps>;
  bodyActions: ITagsActions<BodyProps>;
  htmlActions: ITagsActions<HtmlProps>;
}

export const HelmetScopedContext = createContext<IHelmetScopedContextData | undefined>(undefined);

export const useScopedHelmetContext = () => {
  const context = useContext(HelmetScopedContext);
  if (!context) {
    throw new Error("HelmetScoped context cannot be null, please put tags inside Helmet component");
  }

  return context;
}