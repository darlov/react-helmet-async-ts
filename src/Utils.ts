import { Dispatch, SetStateAction } from "react";
import { IHelmetInstanceState, IHelmetTags, TagProps } from "./Types";

export const addUniqueItem = <T, K extends T[] | undefined>(items: K, item: T, keySelector?: (item: T) => any): K => {
  if (!items) {
    return [item] as K;
  }

  const exist = keySelector !== undefined 
      ? items.some(m => keySelector(m) === keySelector(item))
      : items.some(m => m === item);
  return exist ? items : [...items, item] as K;
}

export const removeAction = <T, K extends T[] | undefined>(action: Dispatch<SetStateAction<K>>, keySelector?: (item: T) => any) => (item: T) => {
  action(items => {
    return keySelector !== undefined 
        ? items?.filter(m => keySelector(m) !== keySelector(item)) as K
        : items?.filter(m => m !== item) as K
  });
};
export const addAction = <T, K extends T[] | undefined>(action: Dispatch<SetStateAction<K>>, keySelector?: (item: T) => any) => {
  return (item: T) => action(items => addUniqueItem<T, K>(items, item, keySelector));
};


export namespace _ {
  export type TSelector<T> = ((elem: T) => any) | keyof T;

  export const sortBy = <T>(array: T[] | undefined, selector: TSelector<T>) => {
    if (typeof selector === "function") {
      return array?.sort((x, y) => selector(x) > selector(y) ? 1 : selector(x) < selector(y) ? -1 : 0);
    }

    return array?.sort((x, y) => x[selector] > y[selector] ? 1 : x[selector] < y[selector] ? -1 : 0);
  }

  export const clear = <T>(array: T[] | undefined) => {
    if (array !== undefined && array.length > 0) {
      array.splice(0, array.length)
    }
  }

  export const isEmpty = (obj: Record<string, any>) => {
    return Object.keys(obj).length === 0;
  }

  type TArray = any[] | undefined
  export const isEmptyArray = (...obj: TArray[]) => {
    return obj.every(m => m === undefined || m.length === 0)
  }

  type InferKey<T> = T extends Partial<Record<infer K, any>> ? K : never;
  type InferValue<T> = T extends Partial<Record<any, infer V>> ? V : never;

  export const toPairs = <T extends Partial<Record<string, any>>>(obj: T) => {
    return Object.entries(obj) as [InferKey<T>, InferValue<T>][];
  };


  type TValSelReturn<TItem, T extends ((arg: TItem, index: number) => any) | undefined>
    = T extends (arg: TItem, index: number) => infer K ? K
    : T extends undefined ? TItem
      : never;

  type TValueSelector<T> = ((item: T, index: number) => T) | ((item: T, index: number) => any)

  export const groupBy = <T, K extends keyof any, V extends TValSelReturn<T, TFunc>, TFunc extends TValueSelector<T>>(items: T[], keySelector: (item: T, index: number) => K | undefined, valueSelector?: TFunc): Map<K, TValSelReturn<T, TFunc>[]> => {
    return items.reduce((result, item, index) => {
      const key = keySelector(item, index);
      if (key !== undefined) {
        if (!result.has(key)) {
          result.set(key, [])
        }
        result.get(key)!.push(valueSelector && valueSelector(item, index));
        return result
      }
      return result;
    }, new Map<K, TValSelReturn<T, TFunc>[]>);
  }
}

export const mergeTags = <T extends TagProps>(tagProp: keyof IHelmetTags, result: T[], instances: IHelmetInstanceState[], primaryAttributeSelector: (tag: T) => string | undefined) => {
  for (const instance of instances) {
    const inTags = instance[tagProp] as T[] | undefined;
    if (inTags) {
      if (instance.emptyState) {
        _.clear(result);
      } else if (result.length === 0) {
        result.push(...inTags);
      } else {
        const instanceGrouped = _.groupBy(inTags, primaryAttributeSelector, (item, index) => ([item, index] as [T, number]));
        const resultGrouped = _.groupBy(result, primaryAttributeSelector, (item, index) => ([item, index] as [T, number]));

        for (const [attr, instanceTags] of instanceGrouped.entries()) {
          const resultTags = resultGrouped.get(attr);

          if (resultTags === undefined) {
            result.push(...instanceTags.map(([tags]) => tags))
          } else if (instanceTags.length !== resultTags.length) {
            result = result.filter(m => !resultTags.some(([tag]) => m === tag));
            result.push(...instanceTags.map(([tag]) => tag))
          } else {
            for (let i = 0; i < instanceTags.length; i++) {
              const [instanceTag] = instanceTags[i];
              const [resultTag, resultIndex] = resultTags[i];

              result[resultIndex] = {...resultTag, ...instanceTag};
            }
          }
        }
      }
    }
  }
  
  return result;
}

