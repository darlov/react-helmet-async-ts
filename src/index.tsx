export { HelmetContextProvider, useHelmetContext } from "./HelmetProvider"
export { Helmet } from "./Helmet"
export * from "./tags"
export { HELMET_ATTRIBUTE } from "./Types"