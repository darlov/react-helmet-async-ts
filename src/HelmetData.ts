import {
  IHelmetDataContext,
  IHelmetInstanceState,
  IHelmetState, TagPriorityConfig,
  ModifyInstanceCallback, ITagPriorityConfigMap, TagConfigName, TypedTagsProps
} from "./types";
import {_, buildServerState, buildState} from "./utils";
import {DefaultTagPriorityConfig} from "./tagConfiguration";


const buildPriorityMap = (priorityConfig: TagPriorityConfig[]) => {
  const map = new Map<TagConfigName, ITagPriorityConfigMap[]>();
  for (let i = 0; i < priorityConfig.length; i++) {
    const config = priorityConfig[i];
    const values = map.get(config.tagName);
    if (values != undefined) {
      values.push({priority: i, config: config})
    } else {
      map.set(config.tagName, [{priority: i, config: config}]);
    }
  }
  return map;
}

const defaultMap = buildPriorityMap(DefaultTagPriorityConfig);

export class HelmetData {
  private _instances = new Map<number, IHelmetInstanceState>();
  private _helmetState: IHelmetState = {tags: [], isEmptyState: false};
  private readonly _priority: Map<TagConfigName, ITagPriorityConfigMap[]>;
  private _stateUpdated = true;

  constructor(priority: TagPriorityConfig[] | undefined, private _context?: IHelmetDataContext, private _canUseDOM = typeof document !== 'undefined') {
    if (priority !== undefined) {
      this._priority = buildPriorityMap(priority);
    } else {
      this._priority = defaultMap;
    }
  }

  get helmetState() {
    this.buildState();
    return this._helmetState;
  }

  get canUseDOM() {
    return this._canUseDOM;
  }

  addInstance = (instance: IHelmetInstanceState) => {
    this._instances.set(instance.id, instance);
    this.triggerUpdates();
  }

  removeInstance = (instance: IHelmetInstanceState) => {
    this._instances.delete(instance.id);
    this.triggerUpdates();
  }

  addItem: ModifyInstanceCallback = (instance, value) => {
    if (instance.tags === undefined) {
      instance.tags = [value as TypedTagsProps]
    } else {
      instance.tags.push(value as TypedTagsProps);
    }

    this.triggerUpdates();
  }

  removeItem: ModifyInstanceCallback = (instance, value) => {
    if (instance.tags !== undefined) {
      const foundIndex = instance.tags.indexOf(value as TypedTagsProps);
      if (foundIndex >= 0) {
        instance.tags.splice(foundIndex, 1);
      }
    }
    this.triggerUpdates();
  }

  private triggerUpdates = () => {
    this._stateUpdated = true;
    if(!this._canUseDOM){
      this.buildState();
    }
  }

  private buildState = () => {
    if (this._stateUpdated) {
      const orderedInstances = _.sortBy([...this._instances.values()], "id");
      this._helmetState = buildState(orderedInstances, this._priority);

      if (!this._canUseDOM && this._context) {
        this._context.state = buildServerState(this._helmetState)
      }

      this._stateUpdated = false;
    }
  }
}