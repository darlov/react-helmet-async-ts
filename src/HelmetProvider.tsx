import { createContext, FC, ReactNode, useContext, useEffect, useMemo, useState } from "react";
import { IHelmetInstanceState } from "./Types";
import { removeAction, addAction } from "./Utils";
import { TagRender } from "./renders";

interface IHelmetContextData {
  addInstance: (state: IHelmetInstanceState) => void;
  removeInstance: (state: IHelmetInstanceState) => void;
}

interface IHelmetContextProviderProps {
  children?: ReactNode
}

const HelmetContext = createContext<IHelmetContextData | undefined>(undefined);

export const HelmetContextProvider: FC<IHelmetContextProviderProps> = ({children}) => {
  const [instances, setInstances] = useState<IHelmetInstanceState[]>();
  const [currentTitle, setCurrentTitle] = useState<HTMLTitleElement | undefined>(() => document.head.getElementsByTagName("title")[0]);

  const context = useMemo<IHelmetContextData>(() => {
    return {
      addInstance: addAction(setInstances, m => m.id),
      removeInstance: removeAction(setInstances, m => m.id),
    }
  }, [setInstances])

  useEffect(() => {
    if (currentTitle && !currentTitle.getAttribute("data-rh")) {
      context.addInstance({
        id: -1,
        emptyState: false,
        titles: [{children: currentTitle.text}]
      })
      currentTitle.remove();
      setCurrentTitle(undefined);
    }
  }, [currentTitle, setCurrentTitle, context.addInstance])


  return (
    <HelmetContext.Provider value={ context }>
      { children }
      <TagRender instances={ instances }/>
    </HelmetContext.Provider>)
}

export const useHelmetContext = () => {
  const context = useContext(HelmetContext);
  if (!context) {
    throw new Error("Helmet context cannot be null, please add a HelmetContextProvider");
  }

  return context;
}


