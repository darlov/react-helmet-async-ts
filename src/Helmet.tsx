import {FC, memo, ReactNode, useEffect, useId, useMemo, useState} from "react";
import {
    createActionsData,
    HelmetScopedContext,
    IHelmetScopedContextData
} from "./HelmetScopedProvider";
import {
    BaseProps, BodyProps,
    IHelmetInstanceState,
    LinkProps,
    MetaProps,
    NoscriptProps,
    ScriptProps,
    StyleProps,
    TitleProps
} from "./Types";
import {_} from "./Utils";
import {useHelmetContext} from "./HelmetProvider";
import {Title} from "./tags";

interface IHelmetProps {
    defaultTitle?: ReactNode,
    children?: ReactNode
}

export const Helmet: FC<IHelmetProps> = memo(({children, defaultTitle}) => {
    const [titles, setTitles] = useState<TitleProps[]>();
    const [metas, setMetas] = useState<MetaProps[]>([]);
    const [styles, setStyles] = useState<StyleProps[]>([]);
    const [scripts, setScripts] = useState<ScriptProps[]>([]);
    const [links, setLinks] = useState<LinkProps[]>([]);
    const [noscripts, setNoscripts] = useState<NoscriptProps[]>([]);
    const [bases, setBases] = useState<BaseProps[]>([]);
    const [bodies, setBodies] = useState<BodyProps[]>([]);
    const rootContext = useHelmetContext();
    const sourceId = useId();
    const id = useMemo(() => {
        return parseInt(sourceId.replace(":r", "").replace(":", ""), 32);
    }, [sourceId])

    const instanceState = useMemo<IHelmetInstanceState | undefined>(() => {
        return {
            id,
            emptyState: _.isEmptyArray(titles, metas, styles, scripts, links, noscripts, bases, bodies),
            titles,
            metas,
            styles,
            scripts,
            links,
            noscripts,
            bases,
            bodies
        }
    }, [id, titles, metas, styles, scripts, links, noscripts, bases, bodies])

    const context = useMemo<IHelmetScopedContextData>(() => {
        return {
            titleActions: createActionsData(setTitles),
            metaActions: createActionsData(setMetas),
            styleActions: createActionsData(setStyles),
            scriptActions: createActionsData(setScripts),
            linkActions: createActionsData(setLinks),
            noscriptActions: createActionsData(setNoscripts),
            baseActions: createActionsData(setBases),
            bodyActions: createActionsData(setBodies)
        }
    }, [setTitles, setMetas, setStyles, setScripts, setLinks, setBases, setBodies]);

    useEffect(() => {
        if (instanceState !== undefined) {
            rootContext.addInstance(instanceState);
            return () => rootContext.removeInstance(instanceState);
        }
    }, [instanceState, rootContext.addInstance, rootContext.removeInstance])

    return (
        <HelmetScopedContext.Provider value={context}>
            {defaultTitle && <Title>{defaultTitle}</Title>}
            {children}
        </HelmetScopedContext.Provider>);
})

Helmet.displayName = "Helmet";