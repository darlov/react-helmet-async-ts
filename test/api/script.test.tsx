import { Helmet, HELMET_ATTRIBUTE, Script } from "../../src";
import { customRender } from "./utils";

describe("script tags", () => {
  describe("Declarative API", () => {
    it("updates script tags", () => {
      const scriptInnerHTML = `
          {
            "@context": "http://schema.org",
            "@type": "NewsArticle",
            "url": "http://localhost/helmet"
          }
        `;
      customRender(
        <Helmet>
          <Script src="http://localhost/test.js" type="text/javascript" />
          <Script src="http://localhost/test2.js" type="text/javascript" />
          <Script type="application/ld+json">{scriptInnerHTML}</Script>
        </Helmet>
      );

      const existingTags = document.head.getElementsByTagName("script");

      expect(existingTags).toBeDefined();

      const filteredTags = [...existingTags].filter(
        tag =>
          (tag.getAttribute("src") === "http://localhost/test.js" &&
            tag.getAttribute("type") === "text/javascript") ||
          (tag.getAttribute("src") === "http://localhost/test2.js" &&
            tag.getAttribute("type") === "text/javascript") ||
          (tag.getAttribute("type") === "application/ld+json" && tag.innerHTML === scriptInnerHTML)
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(3);
    });

    it("clears all scripts tags if none are specified", () => {
      customRender(
        <Helmet>
          <Script src="http://localhost/test.js" type="text/javascript" />
        </Helmet>,
        <Helmet />
      );

      const existingTags = document.head.querySelectorAll(`script[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags without 'src' are not accepted", () => {
      customRender(
        <Helmet>
          <Script property="won't work" />
        </Helmet>
      );

      const existingTags = document.head.querySelectorAll(`script[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("sets script tags based on deepest nested component", () => {
      customRender(
        <div>
          <Helmet>
            <Script src="http://localhost/test.js" type="text/javascript" />
            <Script src="http://localhost/test2.js" type="text/javascript" />
          </Helmet>
        </div>
      );

      const tagNodes = document.head.querySelectorAll(`script[${HELMET_ATTRIBUTE}]`);
      const existingTags = [...tagNodes];
      const firstTag = existingTags[0];
      const secondTag = existingTags[1];

      expect(existingTags).toBeDefined();
      expect(existingTags.length).toBeGreaterThanOrEqual(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag.getAttribute("src")).toBe("http://localhost/test.js");
      expect(firstTag.getAttribute("type")).toBe("text/javascript");
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag.getAttribute("src")).toBe("http://localhost/test2.js");
      expect(secondTag.getAttribute("type")).toBe("text/javascript");
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it("sets undefined attribute values to empty strings", () => {
      customRender(
        <Helmet>
          <Script src="foo.js" async={undefined} />
        </Helmet>
      );

      const existingTag = document.head.querySelector<HTMLScriptElement>(
        `script[${HELMET_ATTRIBUTE}]`
      );

      expect(existingTag).toBeDefined();
      expect(existingTag!.outerHTML).toMatchSnapshot();
    });

    it("does not render tag when primary attribute (src) is null", () => {
      customRender(
        <Helmet>
          <Script src={undefined} type="text/javascript" />
        </Helmet>
      );

      const existingTags = document.head.querySelectorAll<HTMLScriptElement>(
        `script[${HELMET_ATTRIBUTE}]`
      );

      expect(existingTags).toHaveLength(0);
    });

    it("does not render tag when primary attribute (innerHTML) is null", () => {
      customRender(
        <Helmet>
          <Script>{undefined}</Script>
        </Helmet>
      );

      const existingTags = document.head.querySelectorAll<HTMLScriptElement>(
        `script[${HELMET_ATTRIBUTE}]`
      );

      expect(existingTags).toHaveLength(0);
    });
  });
});
