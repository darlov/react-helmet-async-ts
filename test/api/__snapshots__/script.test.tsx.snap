// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`script tags > Declarative API > sets script tags based on deepest nested component 1`] = `"<script src="http://localhost/test.js" type="text/javascript" data-rh="true"></script>"`;

exports[`script tags > Declarative API > sets script tags based on deepest nested component 2`] = `"<script src="http://localhost/test2.js" type="text/javascript" data-rh="true"></script>"`;

exports[`script tags > Declarative API > sets undefined attribute values to empty strings 1`] = `"<script src="foo.js" data-rh="true"></script>"`;
